src/ripple/app/misc/Validations.cpp

```diff
@@ -104,39 +104,67 @@ class ValidationsImp : public Validations
         }
 
         auto hash = val->getLedgerHash ();
-        auto node = val->getNodeID ();
 
-        if (val->isTrusted () && isCurrent)
+        if (isCurrent &&
+            (val->isTrusted () || app_.validators ().listed (signer)))
         {
             ScopedLockType sl (mLock);
 
-            if (!findCreateSet (hash)->insert (std::make_pair (node, val)).second)
+            auto const masterKey = app_.manifestCache ().getMasterKey (signer);
```

Comment:
You're right. There is a race.
I'm working on solution. One thing that definitely needs to change is storing master public keys on the "trusted" list instead of ephemeral signing keys. Otherwise, we'll continue trusting a signing key until the list gets reset, even if the key gets revoked.